File: Coding_0929 Two_Pointers_And_Sliding_Window/이강훈/최원준/3번.py
# https://www.acmicpc.net/problem/1644

'''
1. 아이디어 :
    슬라이딩 윈도우를 사용한다
2. 시간복잡도 :
    O(n)
3. 자료구조 :
    배열
'''


def prime_list(start, end):
    sieve = [True] * end
    if start <= 0:
        start = 2
    m = int(end ** 0.5)
    for i in range(2, m + 1):
        if sieve[i]:
            for j in range(i + i, end, i):
                sieve[j] = False
    return [i for i in range(start, end) if sieve[i]]


n = int(input())
primes = prime_list(2, n + 1) + [0]

l, r, count = 0, 0, 0
total = primes[0]
while r < len(primes) - 1:
    if total < n:
        r += 1
        total += primes[r]
    elif total > n:
        total -= primes[l]
        l += 1
    else:
        r += 1
        total += primes[r]
        total -= primes[l]
        l += 1
        count += 1
print(count)

File: Coding_0929 Two_Pointers_And_Sliding_Window/이강훈/이강훈/3번.java
/*
https://www.acmicpc.net/problem/1644

아이디어
투 포인터

자료구조
ArrayList

시간복잡도
O(N log N) // 에라토스테네스의 체 소수 찾기 시간복잡도,, 소수 배수들을 모두 제거, 중복된 계산은 배제하여 소수 찾기
*/

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        boolean[] isPrime = new boolean[N + 1];
        for (int i = 2; i <= N; i++) {
            isPrime[i] = true;
        }

        for (int i = 2; i * i <= N; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= N; j += i) {
                    isPrime[j] = false;
                }
            }
        }

        List<Integer> primes = new ArrayList<>();
        for (int i = 2; i <= N; i++) {
            if (isPrime[i]) {
                primes.add(i);
            }
        }

        int L = 0;
        int R = 0;
        int total = 0;
        int answer = 0;
        while (true) {
            if (total < N) {
                if (R >= primes.size()) break;

                total += primes.get(R);
                R++;
            } else if (total > N) {
                total -= primes.get(L);
                L++;
            } else {
                answer++;

                total -= primes.get(L);
                L++;
            }
        }

        System.out.println(answer);
    }
}
